prov/udp/src/udpx_init.c

```diff
@@ -38,12 +38,36 @@
 
 int udpx_check_info(struct fi_info *info)
 {
+	if (info && info->ep_attr) {
+		if (info->ep_attr->tx_ctx_cnt == FI_SHARED_CONTEXT) {
+			FI_INFO(&udpx_prov, FI_LOG_CORE,
+					"Shared tx context not supported\n");
+			return -FI_ENODATA;
+		}
+		if (info->ep_attr->rx_ctx_cnt == FI_SHARED_CONTEXT) {
+			FI_INFO(&udpx_prov, FI_LOG_CORE,
+					"Shared rx context not supported\n");
+			return -FI_ENODATA;
+		}
+	}
 	return fi_check_info(&udpx_prov, &udpx_info, info, FI_MATCH_EXACT);
 }
 
 static int udpx_getinfo(uint32_t version, const char *node, const char *service,
 			uint64_t flags, struct fi_info *hints, struct fi_info **info)
 {
+	if (hints && hints->ep_attr) {
+		if (hints->ep_attr->tx_ctx_cnt == FI_SHARED_CONTEXT) {
+			FI_INFO(&udpx_prov, FI_LOG_CORE,
+					"Shared tx context not supported\n");
+			return -FI_ENODATA;
+		}
+		if (hints->ep_attr->rx_ctx_cnt == FI_SHARED_CONTEXT) {
+			FI_INFO(&udpx_prov, FI_LOG_CORE,
+					"Shared rx context not supported\n");
+			return -FI_ENODATA;
+		}
+	}
```

Comment:
I don't like that this check ends up being duplicated.  And it will need to be duplicated for every provider that doesn't support shared contexts.  We need to find a way to push this check into fi_check_info / fi_check_ep_attr.

Similar to how fi_check_info takes 'match_type' as an extra arg, maybe we need some provider specific meta-data, so that fi_check_info can perform more complicated checks.  Maybe we need to change match_type into a set of flags, or pass in a structure with the meta-data instead?